lib/firenest/replicated_state/remote.ex

```diff
@@ -0,0 +1,219 @@
+      # Reconnection, try to catch up
+      %{^ref => old_clock} when clock > old_clock ->
+        {:catch_up, old_clock, state}
+
+      # Reconnection, no remote changes
+      %{^ref => old_clock} ->
+        # Assert for sanity
+        true = old_clock == clock
+        {:ok, state}
+
+      # New node, no state
+      %{} when clock == 0 ->
+        {:ok, %{state | clocks: Map.put(clocks, ref, clock)}}
+
+      # New node, catch up
+      %{} ->
+        {:catch_up, 0, state}
+    end
+  end
+
+  # TODO: Right now down means permdown
+  def down(state, ref) do
+    permdown(state, ref)
+  end
+
+  def permdown(%__MODULE__{clocks: clocks} = state, ref) do
+    true = Map.has_key?(clocks, ref)
+    clocks = Map.delete(clocks, ref)
+    {:delete, ref, %{state | clocks: clocks}}
+  end
+
+  def catch_up(%__MODULE__{clock: current} = state, old_clock, state_getter)
+      when old_clock < current do
+    %{deltas: deltas, tag: tag} = state
+
+    if Map.has_key?(deltas, old_clock) do
+      {:deltas, tag, current, Enum.flat_map(old_clock..current, &Map.fetch!(deltas, &1))}
+    else
+      {:state_transfer, tag, current, state_getter.()}
+    end
+  end
+
+  def broadcast(%__MODULE__{} = state, prepare_delta) do
+    %{pending: pending, clock: clock, tag: tag, broadcast: broadcast, deltas: deltas} = state
+    {:scheduled, broadcast} = broadcast
+
+    new_deltas = prepare_deltas(tag, pending, prepare_delta)
+
+    new_clock = clock + 1
+    deltas = store_deltas(deltas, new_clock, new_deltas)
+    new_state = %{state | pending: %{}, clock: new_clock, broadcast: broadcast, deltas: deltas}
+
+    {{tag, clock, new_deltas}, new_state}
+  end
+
+  def handle_catch_up(%__MODULE__{tag: tag} = state, ref, {:deltas, tag, clock, deltas}) do
+    %{clocks: clocks} = state
+
+    state = %{state | clocks: %{clocks | ref => clock}}
+    {puts, updates, deletes} = handle_deltas(tag, deltas)
+    {:diff, puts, updates, deletes, state}
+  end
+
+  def handle_catch_up(%__MODULE__{tag: tag} = state, from, {:state_transfer, tag, clock, data}) do
+    %{clocks: clocks} = state
+
+    case tag do
+      :ignore -> {:insert, data, %{state | clocks: %{clocks | from => clock}}}
+    end
+  end
+
+  # TODO: should we store somewhere we're catching up with the server?
+  # if so, then we should accumulate the broadcasts we can't handle, until we can.
+  def handle_broadcast(%__MODULE__{clocks: clocks, tag: tag} = state, ref, {tag, clock, delta}) do
+    case clocks do
+      %{^ref => old_clock} when old_clock + 1 == clock ->
+        state = %{state | clocks: %{clocks | ref => clock}}
+        {puts, updates, deletes} = handle_deltas(tag, [delta])
+        {:diff, puts, updates, deletes, state}
+
+      # We missed some broadcast, catch up with the node
+      %{^ref => old_clock} when clock > old_clock ->
+        {:catch_up, old_clock, state}
+
+      # We were caught up with a newer clock than the current, ignore
+      # TODO: is that even possible?
```

Comment:
It is possible if we are getting information from different nodes instead of the node itself. So one node can see A at clock 11 but for the other node B is at clock 13.

Reply:
The topology/synced_server tells us where the information is coming from - we rely only on that to know which node is communicating with us. We don't tag the messages ourselves with the node name.

Reply:
Ok, I think we are agreeing. What I was saying is that, if you can receive the clock for node A from node B, then this can happen. But if you always receive the clock for node A from node A, then this CANNOT happen.